go-controller/pkg/ovn/master.go

```diff
@@ -844,28 +850,64 @@ func (oc *Controller) addNodeAnnotations(node *kapi.Node, hostSubnets []*net.IPN
 
 func (oc *Controller) addNode(node *kapi.Node) ([]*net.IPNet, error) {
 	oc.clearInitialNodeNetworkUnavailableCondition(node, nil)
-
-	hostSubnets, _ := util.ParseNodeHostSubnetAnnotation(node)
-	if hostSubnets != nil {
-		// Node already has subnet assigned; ensure its logical network is set up
-		return hostSubnets, oc.ensureNodeLogicalNetwork(node.Name, hostSubnets)
-	}
-
-	// Node doesn't have a subnet assigned; reserve a new one for it
-	hostSubnets, err := oc.masterSubnetAllocator.AllocateNetworks()
+	hostSubnets, err := util.ParseNodeHostSubnetAnnotation(node)
 	if err != nil {
-		return nil, fmt.Errorf("error allocating network for node %s: %v", node.Name, err)
-	}
-	klog.Infof("Allocated node %s HostSubnet %s", node.Name, util.JoinIPNets(hostSubnets, ","))
+		// Log the error and try to allocate new subnets
+		klog.Infof("Failed to get node %s host subnets annotations: %v", node.Name, err)
+	}
+
+	// OVN can work in single-stack or dual-stack only.
+	expectedHostSubnets := 1
+	// if dual-stack mode we expect one subnet per each IP family
+	if config.IPv4Mode && config.IPv6Mode {
+		expectedHostSubnets = 2
+	}
+	if expectedHostSubnets < len(hostSubnets) {
+		klog.Fatalf("Incompatible number of subnets on node %s: %v subnets assigned but only %d subnets allowed",
+			node.Name, hostSubnets, expectedHostSubnets)
```

Comment:
We will eventually have to support reverting back to single-stack from dual-stack for people who tried it and then changed their mind (or ran into bugs). You can just remove the extra subnet...

Reply:
done in a separate commit https://github.com/ovn-org/ovn-kubernetes/pull/2013/commits/5c4e52df39a12beeb77b9aacedbc23d5cc627570

Reply:
did a bigger refactor to have only one logic to process adding and removing host subnets, so I ammended the changes to the initial commit